Resstock 2024 metadata

mikivee opened the pull request:
## Description
Current surrogate model is trained on 2022.1 metadata, and we want to update so that the model can on both, where old metadata will be mapped to 2024.2. We then want to retrain the model adding in:
* 0.01: 2024.2 baseline 
* 11.07: 2024.2 Med Eff HP Hers no setback
* 13.02: Medium efficiency heat pump with electric back up and no setpoint setback (HERS sizing) and light touch air sealing

See [ticket](https://rewiringamerica.atlassian.net/browse/RAT-37?atlOrigin=eyJpIjoiODlmYTM5YWU4ZGZlNGE2OGIxYjc1ZTZkOTE2M2U3YjgiLCJwIjoiaiJ9) for details of the individual steps that were taken, which will help when reviewing this PR

## Other Notes
* There is a PR out in dmlutils ot update the apply logic test case [here](https://github.com/rewiringamerica/dml-utils/pull/99) and the test cases won't pass until that is merged in
* I will be taking out another small PR to set up better comparison between the modeling results of two versions. For now, you can simply look at the metrics written out in the artifacts folder and look at how the performance change. Broadly, the performance is pretty similar for all the upgrades that the previous model was trained on too.

mikivee committed:
add pytest to dev deps

mikivee committed:
add pytest to dev deps

Commit:
add versioning to all tables and surrogate model based on poetry version

Commit:
remove bucketed benchmark from evaluation script

Commit:
update to less confusing util folder name

Commit:
add versioned metrics

Commit:
put all artifacts in versioned configs

mikivee committed:
run black

Commit:
add unit test for new transformation function

Commit:
fix import path

Commit:
only import spark functions if in db

mikivee committed:
run black

Commit:
fix megastock and run eval

mikivee committed:
remove pytest from dev packages

mikivee committed:
Merge branch 'RAT-262-Update-surrogate-Model-versioning-configs-and-evaluation' of github.com:rewiringamerica/surrogate_modeling into RAT-262-Update-surrogate-Model-versioning-configs-and-evaluation

mikivee committed:
revert to previous packages

Commit:
remove dmlutils import

mikivee committed:
update feature and upgrade docs

mikivee committed:
Merge branch 'RAT-262-Update-surrogate-Model-versioning-configs-and-evaluation' of github.com:rewiringamerica/surrogate_modeling into RAT-262-Update-surrogate-Model-versioning-configs-and-evaluation

Commit:
write eval results to gcs

Commit:
fix import bug

Commit:
change sample size suffixes to be lower case

Commit:
remove datagen print statment

mikivee committed:
add pre v1.0.0 release notes as a pdf

mikivee committed:
update model metrics with full test run

Commit:
rename main scripts to main order clearer

Commit:
make script numbering a suffix

Commit:
remove dup widget

Commit:
refactor to used processed consumption outputs

Commit:
align metadata

Commit:
update notebook docstring

Commit:
update notebook docstring

mikivee committed:
add details on 2022 to 2024 alignment in readme

Commit:
Modify feature transformation to align with 2024.2 metadata

mikivee committed:
document new features

mikivee committed:
Merge branch 'mev/resstock_2024_metadata' of github.com:rewiringamerica/surrogate_modeling into mev/resstock_2024_metadata

Commit:
update feature transformation

Commit:
apply upgrade logic

Commit:
hard code data params

Commit:
small cleanup

mikivee committed:
resolve merge conflict

Commit:
clean up megastock

Commit:
move temp test data

Commit:
link notebooks

mikivee committed:
run black

Commit:
remove megastock upgrade col

Commit:
remove leftover line from merge conflict

nweires commented on tests/test_data_cleaning_utils.py:
```diff
@@ -0,0 +1,61 @@
+from utils import data_cleaning
+
+# TODO: Remove skips for unit test once we have spark testing working on git:
+# https://www.notion.so/rewiringamerica/Local-Spark-Testing-4aef885e20034c18b1a2fba6c355e82c?pvs=4
```
This link doesn't work for me.

mikivee replied:
this is an old srcupt that I just renamed here. I'll just get rid of that line.

nweires commented on docs/features_upgrades.md:
```diff
@@ -66,9 +69,9 @@ Duct Leakage and Insulation:
+Duct Location:
 
-`has_ducts` (bool): Indicator for whether the unit has ducts.
+`duct_location` (bool): Indicator for whether the unit has ducts. TODO: add this features
```
Is this TODO done?

nweires commented on scripts/megastock/data_prep_01.py:
```diff
@@ -1,29 +1,29 @@
 # Databricks notebook source
 # MAGIC %md
-# MAGIC # Process sampled resstock data (from v3.3.0) into same format as 2022.1 file
+# MAGIC # Process sampled resstock data (from v3.3.0)
```
`scripts/megastock/README.md` also mentions using the 2022 format.

mikivee replied:
whoops yeah i totally forgot to update the readme. Fixed. thanks.

mikivee commented on docs/features_upgrades.md:
```diff
@@ -318,6 +334,25 @@ Additional Heat Pump Detail Columns: The heat pump sizing method is specified in
+| `water_heater_location`       | Provides specific locations of water heater (Attic, Crawlspace, Garage, Heated Basement, etc.) | This column did not exist  | Aligned mostly to 2024 by mapping cases where `water_heater_in_unit="No"` to `"None"`and where `water_heater_in_unit="Yes"` to `"Inside Unit"`, which is a new token |
+| `misc_hot_tub_spa` | Possible Values: `"Electricity"`, `"Natural Gas"`, `"Other Fuel"`     | Possible Values: `"Electric"`, `"Gas"`                              | Aligned to 2024 by updating naming conventions |
+| `misc_pool_heater` |Possible Values:  `"Electricity"`, `"Natural Gas"`, `"Other Fuel"`, `"None"`     |Possible Values: `"Electric"`, `"Gas"`, `"Solar"`, `"None"`    | Aligned to 2024 by updating naming conventions and mapping `"Solar"` in 2022 to `"Other Fuel"` |
+| `cooking_range` | Possible Values include `"Electric Resistance"`, `"Electric Induction"`, `"Gas"` | Possible Values: `"Electric"`, `"Gas"`                              | Aligned to 2024 by renaming `"Electric"` to `"Electric Resistance"` |
```
hmm no the name in resstock is just `cooking_range`-- I rename it to fuel when transforming to features.

nweires replied:
Oh, I see! I was thrown off because `cooking_range` also appears in the first row of this table. But that's because it gets processed twice: once to remove the usage info, and again to remove the resistance/induction part.

mikivee replied:
yes, exactly!

mikivee commented on scripts/extract_data_01.py:
```diff
@@ -36,142 +37,277 @@
+    # Appliance Features:
+    # In 2024.2, the usage portion of appliance features are split into a new column and the column with
+    # the appliance name contains only the appliance fuel/type. Further, the usage percentage for appliances is completely determined by “Usage Level”, so there is no need to store each applaince usage levels as additional features.
+    applaince_combined_type_usage_columns = [
+        "clothes_dryer",
+        "clothes_washer",
+        "cooking_range",
+        "dishwasher",
+        "refrigerator",
+    ]
+    for c in applaince_combined_type_usage_columns:
```
lol you have no idea how often I misspell this word

mikivee commented on src/feature_utils.py:
```diff
@@ -775,6 +800,7 @@ def upgrade_to_hp(
             ),
         )
         .withColumn("has_ducted_heating", F.col("has_ducts"))
+        .withColumn("has_ducted_cooling", F.col("has_ducts"))
```
so on line 542 it is set based on the current cooling system:
```
  .withColumn(  # note there are cases where hvac_has_ducts = True, but cooling system is still ductless
      "has_ducted_cooling",
      F.col("hvac_cooling_type").isin(["Central AC", "Ducted Heat Pump"]),
  )
```
Whereas here, we are basically saying as long as the home has ducts, it will get a ducted heat pump in this upgrade. Does that make sense?

Commit:
address PR comments

mikivee committed:
Apply suggestions from code review

Co-authored-by: Natalie Weires <weires.natalie@gmail.com>

Commit:
changing apply_upgrade test data to be stored in this repo and in gcs

mikivee committed:
run black

mikivee commented:
@nweires : addressed all of your comments and went ahead and implemented the [discussed change](https://github.com/rewiringamerica/dml-utils/pull/99) to just move the test data into this repo and export to GCS. I think this is better for a few reasons:
1. We are keeping all other versioned items in GCS, and @SkittyWitty is working on a more long term plan on how to do all this, but in the meantime, it feels simpler to just use 1 strategy/location for all versioning
2. Storing it locally here and in GCS makes is significantly easier to iterate on changes here since doing this in dmutls requires a PR in a 3rd repo every time i make changes, I always have to import a temp version into databricks before pushing to dmutils anyway, and this allows me to test that logic works in both places without having to wait on a new dmutils release.
3. I think storing stacks on stacks of versioned csvs in dmutils doesn't make a lot of sense. Further, I think this aligns more closely with the strategy suggested by Mindy[ here](https://github.com/rewiringamerica/surrogate_modeling/pull/33#discussion_r1999917728) to store only the most recent versions of artifacts  locally so that we can easily track changes from one version to the next in PR review, but store all versions in GCS. I think this also is better to track  and review the changes to these test cases alongside the code.

Commit:
add back mistakenly dropped weather index
